tests/xua_hw_tests/test_dfu.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2024 XMOS LIMITED.
+# This Software is subject to the terms of the XMOS Public Licence: Version 1.
+from pathlib import Path
+import platform
+import pytest
+import re
+import subprocess
+
+from hardware_test_tools.UaDut import UaDut
+from hardware_test_tools.UaDfuApp import UaDfuApp
+
+
+@pytest.fixture(scope="module")
+def factory_xe():
+    xe_path = Path(__file__).parent / "test_dfu" / "bin" / "factory" / "dfu_test_factory.xe"
+    assert xe_path.exists()
+    return xe_path
+
+
+def create_dfu_bin(xe_path):
+    version_re = r"XTC version: (\d+)\.(\d+)\.(\d+)"
+    ret = subprocess.run(["xcc", "--version"], capture_output=True, text=True, timeout=30)
+
+    match = re.search(version_re, ret.stdout)
+    assert match, f"Unable to get XTC Tools version: stdout={ret.stdout}"
+
+    bin_path = xe_path.parent / f"{xe_path.stem}.bin"
+    cmd = ["xflash", "--factory-version", f'{match.group(1)}.{match.group(2)}', "--upgrade", "1", xe_path, "-o", bin_path]
+    ret = subprocess.run(cmd, text=True, capture_output=True, timeout=30)
+    assert ret.returncode == 0, f"Failed to create DFU binary, cmd {cmd}\nstdout:\n{ret.stdout}\nstderr:\n{ret.stderr}"
+
+    return bin_path
+
+
+@pytest.fixture(scope="module")
+def upgrade_bin():
+    xe_path = Path(__file__).parent / "test_dfu" / "bin" / "upgrade" / "dfu_test_upgrade.xe"
+    assert xe_path.exists()
+
+    bin_path = create_dfu_bin(xe_path)
+    assert bin_path.exists()
+    return bin_path
+
+
+def cfg_list():
+    bin_dir = Path(__file__).parent / "test_dfu" / "bin"
+    if not bin_dir.exists():
+        return []
+    exclude_cfgs = ["factory", "upgrade"]
+    all_cfgs = [dir.stem for dir in bin_dir.iterdir()]
+    return [cfg for cfg in all_cfgs if cfg not in exclude_cfgs]
+
+
+def dfu_app_list():
+    if platform.system() == "Windows":
+        return ["custom"]
+    elif platform.system() == "Darwin":
+        return ["custom", "dfu-util"]
+    else:
+        # Until test subdirectories can be rearranged with XCommon CMake builds, an
+        # empty list needs to be returned here for cases where a Linux agent collects
+        # these tests from the top-level test directory (but would filter them out)
+        return []
+
+
+@pytest.mark.parametrize("cfg", cfg_list())
+@pytest.mark.parametrize("dfu_app", dfu_app_list())
+def test_dfu(pytestconfig, factory_xe, upgrade_bin, cfg, dfu_app):
+    if cfg == "i2s_only" and platform.system() == "Windows":
+        pytest.skip("i2s_only not supported on Windows")
```

Comment:
We should probably raise an issue about this so we remember to re-enable when we get a new driver package